ONSdigital/dp-dimension-extractor
# Rework graceful shutdown of service and include healthcheck endpoint

### What

Gracful shutdown will first stop listening to any consumers groups
before coninuing to close connections to the server, other api's,
producing new messages to kafka before finally closing the consumer
group connection. This will aloow the service to commit an in flight
message on success in the given configurable time out, defaulted to 5
seconds.

The above graceful shutdown uses a new version of the go-ns kafka package (which is backward compatible).

Additionally a healthcheck endpoint has been added to the service and is
implemented with the graceful shutdown procedure.

### How to review

Check service gracefully shuts down in the following steps:
1) Shutdown called by os signal terminate
2) Stop listening to kafka consumer group
3) Gracefully close all outbound connections (server, http requests)
4) Close kafka producer
5) Commit any in flight messages that were successful to consumer group
6) Close consumer group
7) Service has stopped

Tests still pass

### Who can review

Team B
